File: backend/src/modules/audit_log/service.py
"""日志审计模块 - 业务逻辑。

v1.1 变更：
- P0-1：_to_csv 删除 _csv_escape 调用，全部交由 csv.writer 按 RFC 4180
        转义。删除 _csv_escape 函数。
- P1-4：export_audit_logs 的 format 参数重命名为 export_format。
- P2-1：CSV 表头与数据行统一由 csv.writer 处理。

包含：
- write_operation_log：中间件调用的写入函数（独立会话）
- list_audit_logs：分页查询
- get_audit_log_detail：详情
- export_audit_logs：导出 CSV / JSON
"""
import csv
import io
import json
import logging
from datetime import datetime, timezone
from typing import Any, Dict, List, Optional, Tuple

from sqlalchemy import func, or_, select
from sqlalchemy.ext.asyncio import AsyncSession

from src.core.database import AsyncSessionLocal
from src.core.exceptions import PlatformException
from src.modules.audit_log.constants import (
    CSV_BOM,
    EXPORT_FORMATS,
    EXPORT_MAX_ROWS,
)
from src.modules.audit_log.models import AuditLogOperation

logger = logging.getLogger(__name__)


# ---------------------------------------------------------------------------
# 内部工具
# ---------------------------------------------------------------------------
def _utcnow_naive() -> datetime:
    return datetime.now(timezone.utc).replace(tzinfo=None)


def _normalize_datetime(dt: Optional[datetime]) -> Optional[datetime]:
    """将带时区的 datetime 转为 UTC naive，与数据库存储格式一致。"""
    if dt is None:
        return None
    if dt.tzinfo is not None:
        return dt.astimezone(timezone.utc).replace(tzinfo=None)
    return dt


def _serialize_log(log: AuditLogOperation) -> Dict[str, Any]:
    """将 AuditLogOperation ORM 对象序列化为统一字典。"""
    return {
        "id": log.id,
        "user_id": log.user_id,
        "username": log.username,
        "module_id": log.module_id,
        "action": log.action,
        "resource": log.resource,
        "resource_id": log.resource_id,
        "detail": log.detail,
        "ip": log.ip,
        "user_agent": log.user_agent,
        "status": log.status,
        "error_code": log.error_code,
        "request_id": log.request_id,
        "created_at": log.created_at,
    }


def _build_filters(
    *,
    module_id: Optional[str],
    user_id: Optional[int],
    action: Optional[str],
    start_time: Optional[datetime],
    end_time: Optional[datetime],
    keyword: Optional[str],
) -> List[Any]:
    """构造查询条件列表（list_audit_logs 与 export_audit_logs 共用）。"""
    conditions: List[Any] = []
    if module_id:
        conditions.append(AuditLogOperation.module_id == module_id)
    if user_id is not None:
        conditions.append(AuditLogOperation.user_id == user_id)
    if action:
        conditions.append(AuditLogOperation.action == action)
    if start_time:
        conditions.append(AuditLogOperation.created_at >= start_time)
    if end_time:
        conditions.append(AuditLogOperation.created_at <= end_time)
    if keyword:
        like = f"%{keyword}%"
        conditions.append(
            or_(
                AuditLogOperation.detail.ilike(like),
                AuditLogOperation.username.ilike(like),
                AuditLogOperation.resource.ilike(like),
            )
        )
    return conditions


# ---------------------------------------------------------------------------
# 写入（中间件调用）
# ---------------------------------------------------------------------------
async def write_operation_log(
    *,
    user_id: Optional[int] = None,
    username: Optional[str] = None,
    module_id: str,
    action: str,
    resource: Optional[str] = None,
    resource_id: Optional[str] = None,
    detail: Optional[str] = None,
    ip: Optional[str] = None,
    user_agent: Optional[str] = None,
    status: str = "success",
    error_code: Optional[int] = None,
    request_id: Optional[str] = None,
) -> None:
    """
    写入一条操作日志。

    - 使用独立数据库会话（AsyncSessionLocal），不依赖请求会话。
    - 内部捕获所有异常并记录 ERROR 日志，绝不影响主请求流程。
    """
    try:
        async with AsyncSessionLocal() as session:
            log = AuditLogOperation(
                user_id=user_id,
                username=username,
                module_id=module_id,
                action=action,
                resource=resource,
                resource_id=resource_id,
                detail=detail,
                ip=ip,
                user_agent=(user_agent or "")[:255] or None,
                status=status,
                error_code=error_code,
                request_id=request_id,
            )
            session.add(log)
            await session.commit()
    except Exception as exc:
        # 审计日志写入失败绝不能影响主流程
        logger.error("写入审计日志失败：%s", exc)


# ---------------------------------------------------------------------------
# 查询
# ---------------------------------------------------------------------------
async def list_audit_logs(
    db: AsyncSession,
    *,
    page: int,
    page_size: int,
    module_id: Optional[str] = None,
    user_id: Optional[int] = None,
    action: Optional[str] = None,
    start_time: Optional[datetime] = None,
    end_time: Optional[datetime] = None,
    keyword: Optional[str] = None,
) -> Dict[str, Any]:
    """分页查询操作日志。"""
    start_time = _normalize_datetime(start_time)
    end_time = _normalize_datetime(end_time)

    if start_time and end_time and start_time > end_time:
        raise PlatformException(
            code=40001,
            message="start_time 不能晚于 end_time",
            status_code=400,
        )

    conditions = _build_filters(
        module_id=module_id,
        user_id=user_id,
        action=action,
        start_time=start_time,
        end_time=end_time,
        keyword=keyword,
    )

    count_stmt = select(func.count(AuditLogOperation.id))
    if conditions:
        count_stmt = count_stmt.where(*conditions)
    total = (await db.execute(count_stmt)).scalar_one()

    list_stmt = select(AuditLogOperation)
    if conditions:
        list_stmt = list_stmt.where(*conditions)
    list_stmt = (
        list_stmt.order_by(
            AuditLogOperation.created_at.desc(),
            AuditLogOperation.id.desc(),
        )
        .offset((page - 1) * page_size)
        .limit(page_size)
    )
    result = await db.execute(list_stmt)
    logs = list(result.scalars().all())

    return {
        "items": [_serialize_log(log) for log in logs],
        "total": total,
        "page": page,
        "page_size": page_size,
        "pages": (total + page_size - 1) // page_size if page_size > 0 else 0,
    }


async def get_audit_log_detail(
    db: AsyncSession, log_id: int
) -> Dict[str, Any]:
    """获取日志详情，不存在返回 40002。"""
    log = await db.get(AuditLogOperation, log_id)
    if not log:
        raise PlatformException(
            code=40002, message="日志不存在", status_code=404
        )
    return _serialize_log(log)


# ---------------------------------------------------------------------------
# 导出
# ---------------------------------------------------------------------------
async def export_audit_logs(
    db: AsyncSession,
    *,
    export_format: str,
    module_id: Optional[str] = None,
    user_id: Optional[int] = None,
    action: Optional[str] = None,
    start_time: Optional[datetime] = None,
    end_time: Optional[datetime] = None,
    keyword: Optional[str] = None,
) -> Tuple[str, str, str]:
    """
    导出操作日志。

    参数：
        export_format: 导出格式，仅支持 "csv" / "json"（v1.1 P1-4 重命名）

    返回：
        (content, media_type, filename)
    """
    if export_format not in EXPORT_FORMATS:
        raise PlatformException(
            code=40004,
            message=f"导出格式不支持：{export_format}，仅支持 {'/'.join(EXPORT_FORMATS)}",
            status_code=400,
        )

    start_time = _normalize_datetime(start_time)
    end_time = _normalize_datetime(end_time)
    if start_time and end_time and start_time > end_time:
        raise PlatformException(
            code=40001,
            message="start_time 不能晚于 end_time",
            status_code=400,
        )

    conditions = _build_filters(
        module_id=module_id,
        user_id=user_id,
        action=action,
        start_time=start_time,
        end_time=end_time,
        keyword=keyword,
    )

    stmt = select(AuditLogOperation)
    if conditions:
        stmt = stmt.where(*conditions)
    stmt = (
        stmt.order_by(
            AuditLogOperation.created_at.desc(),
            AuditLogOperation.id.desc(),
        )
        .limit(EXPORT_MAX_ROWS)
    )
    result = await db.execute(stmt)
    logs = list(result.scalars().all())

    timestamp = _utcnow_naive().strftime("%Y%m%d_%H%M%S")

    try:
        if export_format == "csv":
            content = _to_csv(logs)
            media_type = "text/csv; charset=utf-8"
            filename = f"audit_logs_{timestamp}.csv"
        else:  # json
            content = _to_json(logs)
            media_type = "application/json; charset=utf-8"
            filename = f"audit_logs_{timestamp}.json"
    except Exception as exc:
        raise PlatformException(
            code=40005,
            message=f"日志导出失败：{exc}",
            status_code=500,
        ) from exc

    return content, media_type, filename


def _to_csv(logs: List[AuditLogOperation]) -> str:
    """
    生成 CSV 内容（含 UTF-8 BOM）。

    v1.1 修复 P0-1：
        - 不再手动调用 _csv_escape 转义字段；
        - 全部交由 csv.writer 按 RFC 4180 处理；
        - 避免双重转义：三层引号会破坏数据，正确形式应为单层双引号包裹。

    v1.1 修复 P2-1：
        - 表头与数据行统一由 csv.writer 处理，风格一致。

    示例（RFC 4180 转义规则由 csv.writer 实现）：
        输入 a,b           → 输出 "a,b"
        输入 a"b           → 输出 "a""b"
        输入 a 换行 b      → 输出 "a\nb"（字段被双引号包裹）
    """
    buf = io.StringIO()
    buf.write(CSV_BOM)
    writer = csv.writer(buf, lineterminator="\n")
    # 表头：同样交由 csv.writer 处理
    writer.writerow(
        [
            "id",
            "user_id",
            "username",
            "module_id",
            "action",
            "resource",
            "resource_id",
            "detail",
            "ip",
            "user_agent",
            "status",
            "error_code",
            "request_id",
            "created_at",
        ]
    )
    for log in logs:
        # P0-1：直接传原始值，由 csv.writer 统一按 RFC 4180 转义
        writer.writerow(
            [
                log.id,
                log.user_id if log.user_id is not None else "",
                log.username if log.username is not None else "",
                log.module_id if log.module_id is not None else "",
                log.action if log.action is not None else "",
                log.resource if log.resource is not None else "",
                log.resource_id if log.resource_id is not None else "",
                log.detail if log.detail is not None else "",
                log.ip if log.ip is not None else "",
                log.user_agent if log.user_agent is not None else "",
                log.status if log.status is not None else "",
                log.error_code if log.error_code is not None else "",
                log.request_id if log.request_id is not None else "",
                log.created_at.isoformat() if log.created_at else "",
            ]
        )
    return buf.getvalue()


def _to_json(logs: List[AuditLogOperation]) -> str:
    """生成 JSON 内容。"""
    items = [_serialize_log(log) for log in logs]
    for item in items:
        if item.get("created_at") is not None:
            item["created_at"] = item["created_at"].isoformat()
    return json.dumps(
        {"total": len(items), "items": items},
        ensure_ascii=False,
        indent=2,
    )

File: backend-go/internal/modules/audit_log/service.go
package audit_log
import (
	"bytes"
	"context"
	"encoding/csv"
	"encoding/json"
	"errors"
	"fmt"
	"strconv"
	"strings"
	"sync"
	"time"
	"github.com/rs/zerolog/log"
	"gorm.io/gorm"
	"backend-go/internal/config"
	"backend-go/internal/exception"
	"backend-go/internal/models"
)
// Service 日志审计服务。
type Service struct {
	db  *gorm.DB
	cfg *config.Config
	wg  sync.WaitGroup
}
// NewService 创建日志审计服务。
func NewService(db *gorm.DB, cfg *config.Config) *Service {
	return &Service{db: db, cfg: cfg}
}
// Wait 等待所有异步写入任务完成（供优雅关闭调用）。
func (s *Service) Wait() {
	s.wg.Wait()
}
// WriteOperationLogAsync 异步写入一条操作日志。
//
// 调用方负责将空字符串字段归一化为 nil（通过 strPtrOrNil）。
func (s *Service) WriteOperationLogAsync(entry *models.AuditLogOperation) {
	s.wg.Add(1)
	go func() {
		defer s.wg.Done()
		ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
		defer cancel()
		if err := s.db.WithContext(ctx).Create(entry).Error; err != nil {
			log.Error().Err(err).Msg("写入审计日志失败")
		}
	}()
}
// WriteOperationLog 同步写入一条操作日志（供测试使用）。
func (s *Service) WriteOperationLog(entry *models.AuditLogOperation) error {
	return s.db.Create(entry).Error
}
// ListAuditLogs 分页查询操作日志。
//
// v1.1（P1-3）：Count 后使用 Session(&gorm.Session{}) 隔离，避免
// Statement 污染导致的 Order / Limit 失效。
//
// v1.2（P2-NEW-1）：分页参数从 Handler 层传入的非 nil 值展开。
func (s *Service) ListAuditLogs(q AuditLogListQuery) (map[string]interface{}, error) {
	page := 1
	if q.Page != nil {
		page = *q.Page
	}
	pageSize := 20
	if q.PageSize != nil {
		pageSize = *q.PageSize
	}
	startTime, err := ParseTime(q.StartTime)
	if err != nil {
		return nil, exception.New(exception.CodeValidationFail, "请求参数校验失败", 422,
			map[string]interface{}{
				"errors": []map[string]interface{}{
					{
						"type": "datetime_parsing",
						"loc":  []string{"query", "start_time"},
						"msg":  fmt.Sprintf("时间格式无效：%s", q.StartTime),
					},
				},
			})
	}
	endTime, err := ParseTime(q.EndTime)
	if err != nil {
		return nil, exception.New(exception.CodeValidationFail, "请求参数校验失败", 422,
			map[string]interface{}{
				"errors": []map[string]interface{}{
					{
						"type": "datetime_parsing",
						"loc":  []string{"query", "end_time"},
						"msg":  fmt.Sprintf("时间格式无效：%s", q.EndTime),
					},
				},
			})
	}
	if startTime != nil && endTime != nil && startTime.After(*endTime) {
		return nil, exception.New(exception.CodeAuditLogInvalidTimeRange,
			"start_time 不能晚于 end_time", 400, nil)
	}
	query := s.db.Model(&models.AuditLogOperation{})
	if q.ModuleID != "" {
		query = query.Where("module_id = ?", q.ModuleID)
	}
	if q.UserID != nil {
		query = query.Where("user_id = ?", *q.UserID)
	}
	if q.Action != "" {
		query = query.Where("action = ?", q.Action)
	}
	if startTime != nil {
		query = query.Where("created_at >= ?", *startTime)
	}
	if endTime != nil {
		query = query.Where("created_at <= ?", *endTime)
	}
	if q.Keyword != "" {
		like := "%" + q.Keyword + "%"
		query = query.Where(
			"detail LIKE ? OR username LIKE ? OR resource LIKE ?",
			like, like, like,
		)
	}
	var total int64
	if err := query.Count(&total).Error; err != nil {
		return nil, exception.New(exception.CodeInternalError, "查询日志失败", 500, nil)
	}
	var logs []models.AuditLogOperation
	// v1.1（P1-3）：Session 隔离，避免 Count 修改 Statement 影响 Find
	if err := query.Session(&gorm.Session{}).
		Order("created_at DESC, id DESC").
		Offset((page - 1) * pageSize).
		Limit(pageSize).
		Find(&logs).Error; err != nil {
		return nil, exception.New(exception.CodeInternalError, "查询日志失败", 500, nil)
	}
	items := make([]AuditLogOut, 0, len(logs))
	for i := range logs {
		items = append(items, *serializeAuditLog(&logs[i]))
	}
	pages := 0
	if pageSize > 0 {
		pages = int((total + int64(pageSize) - 1) / int64(pageSize))
	}
	return map[string]interface{}{
		"items":     items,
		"total":     total,
		"page":      page,
		"page_size": pageSize,
		"pages":     pages,
	}, nil
}
// GetAuditLogDetail 获取日志详情。
func (s *Service) GetAuditLogDetail(logID uint) (*AuditLogOut, error) {
	var entry models.AuditLogOperation
	if err := s.db.First(&entry, logID).Error; err != nil {
		if errors.Is(err, gorm.ErrRecordNotFound) {
			return nil, exception.New(exception.CodeAuditLogNotFound, "日志不存在", 404, nil)
		}
		return nil, exception.New(exception.CodeInternalError, "查询日志失败", 500, nil)
	}
	return serializeAuditLog(&entry), nil
}
// ExportAuditLogs 导出操作日志。
// 返回：内容、MediaType、文件名。
func (s *Service) ExportAuditLogs(q AuditLogExportQuery) ([]byte, string, string, error) {
	// v1.1（P2-4）：统一小写
	format := strings.ToLower(q.Format)
	if format != ExportFormatCSV && format != ExportFormatJSON {
		return nil, "", "", exception.New(
			exception.CodeAuditLogExportFormat,
			fmt.Sprintf("导出格式不支持：%s，仅支持 csv/json", q.Format),
			400,
			nil,
		)
	}
	startTime, err := ParseTime(q.StartTime)
	if err != nil {
		return nil, "", "", exception.New(exception.CodeValidationFail, "请求参数校验失败", 422,
			map[string]interface{}{
				"errors": []map[string]interface{}{
					{
						"type": "datetime_parsing",
						"loc":  []string{"query", "start_time"},
						"msg":  fmt.Sprintf("时间格式无效：%s", q.StartTime),
					},
				},
			})
	}
	endTime, err := ParseTime(q.EndTime)
	if err != nil {
		return nil, "", "", exception.New(exception.CodeValidationFail, "请求参数校验失败", 422,
			map[string]interface{}{
				"errors": []map[string]interface{}{
					{
						"type": "datetime_parsing",
						"loc":  []string{"query", "end_time"},
						"msg":  fmt.Sprintf("时间格式无效：%s", q.EndTime),
					},
				},
			})
	}
	if startTime != nil && endTime != nil && startTime.After(*endTime) {
		return nil, "", "", exception.New(exception.CodeAuditLogInvalidTimeRange,
			"start_time 不能晚于 end_time", 400, nil)
	}
	query := s.db.Model(&models.AuditLogOperation{})
	if q.ModuleID != "" {
		query = query.Where("module_id = ?", q.ModuleID)
	}
	if q.UserID != nil {
		query = query.Where("user_id = ?", *q.UserID)
	}
	if q.Action != "" {
		query = query.Where("action = ?", q.Action)
	}
	if startTime != nil {
		query = query.Where("created_at >= ?", *startTime)
	}
	if endTime != nil {
		query = query.Where("created_at <= ?", *endTime)
	}
	if q.Keyword != "" {
		like := "%" + q.Keyword + "%"
		query = query.Where(
			"detail LIKE ? OR username LIKE ? OR resource LIKE ?",
			like, like, like,
		)
	}
	var logs []models.AuditLogOperation
	if err := query.Session(&gorm.Session{}).
		Order("created_at DESC, id DESC").
		Limit(ExportMaxRows).
		Find(&logs).Error; err != nil {
		return nil, "", "", exception.New(exception.CodeAuditLogExportFailed, "日志导出失败", 500, nil)
	}
	timestamp := time.Now().UTC().Format("20060102_150405")
	if format == ExportFormatCSV {
		content, err := toCSV(logs)
		if err != nil {
			return nil, "", "", exception.New(
				exception.CodeAuditLogExportFailed,
				fmt.Sprintf("日志导出失败：%v", err),
				500, nil,
			)
		}
		return content, "text/csv; charset=utf-8", fmt.Sprintf("audit_logs_%s.csv", timestamp), nil
	}
	content, err := toJSON(logs)
	if err != nil {
		return nil, "", "", exception.New(
			exception.CodeAuditLogExportFailed,
			fmt.Sprintf("日志导出失败：%v", err),
			500, nil,
		)
	}
	return content, "application/json; charset=utf-8", fmt.Sprintf("audit_logs_%s.json", timestamp), nil
}
// ParseTime 解析时间字符串，支持 RFC3339 与无时区 ISO 格式。
func ParseTime(s string) (*time.Time, error) {
	if s == "" {
		return nil, nil
	}
	layouts := []string{
		time.RFC3339,
		"2006-01-02T15:04:05",
		"2006-01-02 15:04:05",
	}
	for _, layout := range layouts {
		if t, err := time.Parse(layout, s); err == nil {
			t = t.UTC()
			return &t, nil
		}
	}
	return nil, fmt.Errorf("invalid time format: %s", s)
}
// serializeAuditLog 将 ORM 对象序列化为响应结构。
func serializeAuditLog(entry *models.AuditLogOperation) *AuditLogOut {
	return &AuditLogOut{
		ID:         entry.ID,
		UserID:     entry.UserID,
		Username:   entry.Username,
		ModuleID:   entry.ModuleID,
		Action:     entry.Action,
		Resource:   entry.Resource,
		ResourceID: entry.ResourceID,
		Detail:     entry.Detail,
		IP:         entry.IP,
		UserAgent:  entry.UserAgent,
		Status:     entry.Status,
		ErrorCode:  entry.ErrorCode,
		RequestID:  entry.RequestID,
		CreatedAt:  formatDateTime(entry.CreatedAt),
	}
}
// toCSV 生成 CSV 内容（含 UTF-8 BOM），全部交由 encoding/csv 按 RFC 4180 转义。
//
// v1.1（P2-2）：逐条检查 w.Write 返回错误。
func toCSV(logs []models.AuditLogOperation) ([]byte, error) {
	buf := new(bytes.Buffer)
	buf.WriteString(CSV_BOM)
	w := csv.NewWriter(buf)
	if err := w.Write([]string{
		"id", "user_id", "username", "module_id", "action", "resource",
		"resource_id", "detail", "ip", "user_agent", "status",
		"error_code", "request_id", "created_at",
	}); err != nil {
		return nil, err
	}
	for _, entry := range logs {
		record := []string{
			strconv.FormatUint(uint64(entry.ID), 10),
			uintPtrToString(entry.UserID),
			strPtrToString(entry.Username),
			entry.ModuleID,
			entry.Action,
			strPtrToString(entry.Resource),
			strPtrToString(entry.ResourceID),
			strPtrToString(entry.Detail),
			strPtrToString(entry.IP),
			strPtrToString(entry.UserAgent),
			entry.Status,
			intPtrToString(entry.ErrorCode),
			strPtrToString(entry.RequestID),
			formatDateTime(entry.CreatedAt),
		}
		if err := w.Write(record); err != nil {
			return nil, err
		}
	}
	w.Flush()
	if err := w.Error(); err != nil {
		return nil, err
	}
	return buf.Bytes(), nil
}
// toJSON 生成 JSON 内容。
func toJSON(logs []models.AuditLogOperation) ([]byte, error) {
	items := make([]AuditLogOut, 0, len(logs))
	for i := range logs {
		items = append(items, *serializeAuditLog(&logs[i]))
	}
	return json.MarshalIndent(map[string]interface{}{
		"total": len(items),
		"items": items,
	}, "", "  ")
}
func strPtrToString(p *string) string {
	if p == nil {
		return ""
	}
	return *p
}
func uintPtrToString(p *uint) string {
	if p == nil {
		return ""
	}
	return strconv.FormatUint(uint64(*p), 10)
}
func intPtrToString(p *int) string {
	if p == nil {
		return ""
	}
	return strconv.Itoa(*p)
}
